src/cpp/session/modules/rmarkdown/SessionRmdNotebook.cpp

```diff
@@ -0,0 +1,670 @@
+   if (error)
+   {
+      LOG_ERROR(error);
+   }
+
+   FilePath outputCsv = chunkConsolePath(
+         module_context::resolveAliasedPath(path).absolutePath(), 
+         docId, chunkId);
+
+   std::vector<std::string> vals; 
+   vals.push_back(safe_convert::numberToString(type));
+   vals.push_back(output);
+   error = core::writeStringToFile(outputCsv, 
+         text::encodeCsvLine(vals) + "\n", 
+         string_utils::LineEndingPassthrough, truncate);
+   if (error)
+   {
+      LOG_ERROR(error);
+   }
+}
+
+void onConsoleOutput(module_context::ConsoleOutputType type, 
+      const std::string& output)
+{
+   if (type == module_context::ConsoleOutputNormal)
+      onConsoleText(s_consoleDocId, s_consoleChunkId, kChunkConsoleOutput, 
+                    output, false);
+   else
+      onConsoleText(s_consoleDocId, s_consoleChunkId, kChunkConsoleError, 
+                    output, false);
+}
+
+void onConsoleInput(const std::string& input)
+{
+   onConsoleText(s_consoleDocId, s_consoleChunkId, kChunkConsoleInput, input, 
+         false);
+}
+
+void disconnectConsole()
+{
+   module_context::events().onConsolePrompt.disconnect(onConsolePrompt);
+   module_context::events().onConsoleOutput.disconnect(onConsoleOutput);
+   module_context::events().onConsoleInput.disconnect(onConsoleInput);
+   s_consoleConnected = false;
+}
+
+void connectConsole()
+{
+   module_context::events().onConsolePrompt.connect(onConsolePrompt);
+   module_context::events().onConsoleOutput.connect(onConsoleOutput);
+   module_context::events().onConsoleInput.connect(onConsoleInput);
+   s_consoleConnected = true;
+}
+
+void onActiveConsoleChanged(const std::string& consoleId)
+{
+   s_activeConsole = consoleId;
+   if (consoleId == s_consoleChunkId)
+   {
+      if (s_consoleConnected) 
+         return;
+      connectConsole();
+   }
+   else if (s_consoleConnected)
+   {
+      // some other console is connected; disconnect ours
+      disconnectConsole();
+   }
+}
+
+Error handleChunkOutputRequest(const http::Request& request,
+                               http::Response* pResponse)
+{
+   // uri format is: /chunk_output/<doc-id>/...
+   
+   // split URI into pieces, extract the document ID, and remove that part of
+   // the URI
+   std::vector<std::string> parts = algorithm::split(request.uri(), "/");
+   if (parts.size() < 4) 
+      return Success();
+   std::string docId = parts[2];
+   for (int i = 0; i < 3; i++)
+      parts.erase(parts.begin());
+
+   std::string path;
+   Error error = source_database::getPath(docId, &path);
+   if (error)
+      return error;
+   FilePath target = chunkCacheFolder(path, docId).complete(
+         algorithm::join(parts, "/"));
+
+   if (parts[0] == kChunkLibDir)
+   {
+      // if a reference to the chunk library folder, we can reuse the contents
+      pResponse->setCacheableFile(target, request);
+   }
+   else
+   {
+      // otherwise, we don't want the output to be cached (so we get fresh
```

Comment:
You could use setBody which could do eTag based caching (so at least we save the transit time if the contents haven't changed).